torch/multiprocessing/reductions.py

```diff
@@ -23,10 +23,108 @@ def __init__(self, ptr):
         self.cdata = ptr
 
 
-# mapping from handles to StorageRef objects
+# mapping from handles to StorageRef objects. Doesn't include CUDA handles.
 shared_cache = weakref.WeakValueDictionary()
 
 
+# CUDA storage caching strategy
+# -----------------------------
+#
+# All CUDA storages are cached into 'cuda_cache'.
+#
+# Call a "CUDA storage that processes wish to share" a "real storage".
+# "real storages" are the storages that other processes want to and
+# eventually will receive.
+#
+# Call a "CUDA storage that points to the base of an allocation block" a
+# "base storage". These are obtained by opening CUDA handles to the
+# base of an allocation block, where multiple "real storages" may live.
+# Those "real storages" can be obtained by adding an offset to "base storage"
+#
+# "base storages" are cached into the "base cache" and "real storages" are
+# cached into the "real cache"
+#
+# When a process shares a cuda storage:
+# - If the cuda storage is from opening a received handle, find the handle
+#   in the metadata cache and send the returned metadata.
+# - If the cuda storage is from allocating storage, get the handle, build
+#   the metadata, and send it.
+#
+# When a process receives a cuda storage:
+# - Check the caches:
+#   - Check the "real cache" for the real storage.
+#   - If it's not there, look in the "base cache".
+#   - If the storage exists in the "base cache", recreate the real storage by
+#     add it to the "base storage"
+# - If it's not in the caches, recreate the real and base storages and cache
+#   both.
+# - In addition, if this process received a handle that it has opened
+#   (either now or in the past), we cache the storage's metadata in the
+#   metadata cache. This is so that the process can access the handle
+#   to send to other processes.
+#   A process that didn't open a handle (the process that allocated the storage)
+#   doesn't cache the metadata.
+#
+# "base storages" are uniquely identified by their CUDA handle.
+# "real storages" are uniquely identified by (handle, offset).
```

Comment:
This doesn't sounds true to me. Why couldn't you have two storages with same base and same offset?

Reply:
Right, we could have `storage1` have `(base, offset, view_size=10)` and `storage2` have `(base, offset, view_size=20)` and then they'd be different